### ISSUE TITLE
rule re-firing incorrectly

### ISSUE THREAD

**Issue by tve** [anywhichway/rule-reactor#11]
This issue may well be the same as #10, but just in case I'm opening a separate issue. When I run rule reactor and subsequently add new facts and re-run rule-reactor I see that old rule activations re-fire. I hope this is a bug and not the way it's supposed to work. Here is a simple example:

```
var RuleReactor = require("rule-reactor");
var reactor = new RuleReactor();

function Person(r) { Object.assign(this, r); }
function Home(r) { Object.assign(this, r); }
global.Person = Person;
global.Home = Home;

LL = console.log;

reactor.createRule("has-a-home", 0, {p: Person, h: Home},
        function(p, h) { return p.home_id == h.id; },
        function(p, h) { LL("person", p.name, "has home", h.id); }
);

reactor.createRule("stop", -100, {},
    function() { return true; },
    function() { reactor.stop(); }
);

reactor.trace(0);

// define objects
for (var i=0; i<4; i++) {
        reactor.assert(new Person({name: "p"+i, home_id: i}));
}
for (var i=0; i<4; i++) {
        reactor.assert(new Home({id: i}));
}

function runReactor(cb) {
        reactor.run(Infinity, false, function() {
                console.log("Data Count: ", reactor.data.size);
                console.log("Executions: ", reactor.run.executions);
                console.log("RPS: ", reactor.run.rps);
                Object.keys(reactor.rules).forEach(function(rulename) {
                        var rule = reactor.rules[rulename];
                        console.log(rule.name, "-- matches:"+rule.potentialMatches,
                                "tested:"+rule.tested, "activated:"+rule.activated, "fired:"+rule.fired);
                });
                if (cb) cb();
        });
}

LL("start");
runReactor(function() {'
        // now add one fact and re-run the reactor
        LL("add");
        reactor.assert(new Home({id: 5}));
        runReactor();
});
```

A run looks as follows:
```
> node test2.js
start
person p3 has home 3
person p2 has home 2
person p1 has home 1
person p0 has home 0
person p0 has home 0
person p0 has home 0
person p0 has home 0
Data Count:  8
Executions:  8
RPS:  1600
has-a-home -- matches:16 tested:4 activated:10 fired:7
stop -- matches:0 tested:1 activated:1 fired:1
add
person p3 has home 3
person p2 has home 2
person p1 has home 1
person p0 has home 0
Data Count:  9
Executions:  5
RPS:  1250
has-a-home -- matches:20 tested:5 activated:14 fired:11
stop -- matches:0 tested:2 activated:2 fired:2
```
I did not expect any rule to fire after the "add".

**Comment by anywhichway** [comment]
This appears to be a bug. Will investigate over the next few days.

**Comment by tve** [comment]
Related to this bug, the semantics of modifying a fact are not actually clear to me at all. I suspect that the theoretical formulation of a rules engine doesn't support fact modification, only assertion and retraction. If I consider a rule R with a condition Rc and an action Ra, plus a fact F1, then I believe it's clear that if I want to modify fact F1 I have to retract F1 and assert F2. When F2 is asserted it is irrelevant whether Rc was evaluated on F1 or Ra was fired on F1. F2 simply gets evaluated as a new stand-alone fact.

If fact modification is allowed (as in rule-reactor) then things are not very clear to me. Suppose I assert F1 with attributes a1 and a2 (F1.a1, F1.a2). Suppose Rc checks F1.a1 and returns true, then Ra is fired on F1. I now modify F1.a1. It is clear that Rc needs to be reevaluated. Suppose it returns true again, then is Ra fired again because this is a new "version" of F1 or is it not fired again because Ra has already been run on F1?

Suppose I now modify F1.a2. Rc does not need to be reevaluated because it does not check F1.a2. But Rc is true on F1. Is Ra fired again because this is a new version of F1 or is it not fired again because Ra has already been run on F1?

Now suppose I modify F1.a1 twice: the first time it renders Rc false, the second time it renders Rc true again. Is Ra fired again on this second modification because the evaluation of Rc transitioned from false to true or is it not fired again because Ra was fired on F1 before and an action only ever gets fired once on any fact?

For my use-case I would not want an action to be fired again if an unrelated field changes. But I have to admit I'm not sure about this last question, i.e., whether it should fire again if the condition transitions to false and then true again.

Thoughts?

**Comment by anywhichway** [comment]
With respect to: "Related to this bug, the semantics of modifying a fact are not actually clear to me at all. I suspect that the theoretical formulation of a rules engine doesn't support fact modification, only assertion and retraction. If I consider a rule R with a condition Rc and an action Ra, plus a fact F1, then I believe it's clear that if I want to modify fact F1 I have to retract F1 and assert F2. When F2 is asserted it is irrelevant whether Rc was evaluated on F1 or Ra was fired on F1. F2 simply gets evaluated as a new stand-alone fact." Most rule engines do support fact modification. In your example above there is no need to retract F1, just assert F2 and the rule will fire again.

With respect to: "If fact modification is allowed (as in rule-reactor) then things are not very clear to me. Suppose I assert F1 with attributes a1 and a2 (F1.a1, F1.a2). Suppose Rc checks F1.a1 and returns true, then Ra is fired on F1. I now modify F1.a1. It is clear that Rc needs to be reevaluated. Suppose it returns true again, then is Ra fired again because this is a new "version" of F1 or is it not fired again because Ra has already been run on F1?" It should be fired again because the rule engine reacts to changes at the attribute level. Think of each object as though it were represented as a series of facts (objectid attribute value), e.g. (1 name "joe"),(1 age 21),(2 name "mary"),(2 age 20). The addition of or changes to facts will should rules to fire, e.g. (1 name "joey"),(3 name "bill"). In fact, the early rule engines this is exactly how object were represented.

With respect to "Suppose I now modify F1.a2. Rc does not need to be reevaluated because it does not check F1.a2. But Rc is true on F1. Is Ra fired again because this is a new version of F1 or is it not fired again because Ra has already been run on F1?" Ra should not be fired again. Ra should only fire if Rc is relevant and returns true.

With respect to: "Now suppose I modify F1.a1 twice: the first time it renders Rc false, the second time it renders Rc true again. Is Ra fired again on this second modification because the evaluation of Rc transitioned from false to true or is it not fired again because Ra was fired on F1 before and an action only ever gets fired once on any fact?" Ra should fire again. Once again, think of the (object attribute values) as facts and the objects themselves simply as convenient containers.

Perhaps a little explanation regarding rule-reactor internals will help:

1) Every rule is "compiled" so that its domain (class) and range (properties that get tested) are known. During this compilation the classes prototypes of domains are modified to note which properties on instances are referenced by any rule at all. These are called "activeKeys". The prototypes are also modified to track which rules reference the "activeKeys".
2) For each object asserted all active keys are are modified using Object.defineProperty so that changes to their values force re-evaluation of the rules relevant to the key modified. If the re-evaluation returns true, then the action is put on an agenda (queue) in the form of an Activation for potential invocation. The Activations keep track of which data combinations they exist to represent so that if the data changes they can be removed from the agenda without the action being invoked.

Based on your questions in this issue and other issues, we are suspicious you have either uncovered a bug or that compilation is occurring incorrectly as a nuance of how you have implemented your classes. We get a sense that despite your self skepticism, you do have a sense of the correct behavior. We will be investigating further over the next few days.

**Comment by tve** [comment]
Do you have any guidance on an ETA towards fixing this issue? I have been blocked for the last week and need to make a decision on how to move forwards. I'm very grateful that you have made rule-reactor available and I fully realize this is a free open-source project, so I'm just asking for guidance. Thanks!

**Comment by anywhichway** [comment]
We are still researching this. Sorry it is taking so long, this is code base is not our primary focus at the moment ... although we are spending time researching your issues. Can you architect you system so that you do not stop and restart? i.e. don't have your stop rule actually stop until you assert a fact "stop", perhaps from an external source?

**Comment by tve** [comment]
This test case is fixed by the patch to #10. I'm not sure that patch fixes all instances of this problem. The take-away is that `rule.test` must only be called on exactly the combinations that truly need to be retested, not fewer but also not any more. That seems fragile.

**Comment by anywhichway** [comment]
fixed courtesy of @tve
